internal/v5/archive_test.go

```diff
@@ -1729,3 +1756,125 @@ func (s *ArchiveSearchSuite) TestGetSearchUpdate(c *gc.C) {
 		stats.CheckSearchTotalDownloads(c, s.store, &url.URL, 1)
 	}
 }
+
+type ArchiveSuiteWithTerms struct {
+	commonArchiveSuite
+}
+
+var _ = gc.Suite(&ArchiveSuiteWithTerms{})
+
+func (s *ArchiveSuiteWithTerms) SetUpSuite(c *gc.C) {
+	s.commonSuite.SetUpSuite(c)
+	s.enableTerms = true
+	s.enableIdentity = true
+}
+
+func (s *ArchiveSuiteWithTerms) SetUpTest(c *gc.C) {
+	s.commonSuite.SetUpTest(c)
+	s.discharge = dischargeForUser("bob")
+}
+
+func (s *ArchiveSuiteWithTerms) TestGetUserHasAgreedToTermsAndConditions(c *gc.C) {
+	s.dischargeTerms = func(s1, s2 string) ([]checkers.Caveat, error) {
+		return nil, nil
+	}
+
+	client := httpbakery.NewHTTPClient()
+
+	id := newResolvedURL("cs:~charmers/precise/terms-0", -1)
+	terms := s.assertUploadCharm(c, "POST", id, "terms")
+	err := s.store.SetPerms(&id.URL, "read", "bob")
+	c.Assert(err, gc.IsNil)
+
+	archiveBytes, err := ioutil.ReadFile(terms.Path)
+	c.Assert(err, gc.IsNil)
+
+	archiveUrl := storeURL("~charmers/precise/terms-0/archive")
+	rec := httptesting.DoRequest(c, httptesting.DoRequestParams{
+		Handler: s.srv,
+		URL:     archiveUrl,
+		Do:      bakeryDo(client),
+	})
+	c.Assert(rec.Code, gc.Equals, http.StatusOK)
+	c.Assert(rec.Body.Bytes(), gc.DeepEquals, archiveBytes)
+	c.Assert(rec.Header().Get(params.ContentHashHeader), gc.Equals, hashOfBytes(archiveBytes))
```

Comment:
I don't think we need to duplicate the whole of TestGet here - the download logic is the same in each case, so we actually only need to check that we can get the archive.
Given that, we can probably delete lines 1800 to 1816 here.